### PR TITLE
Use new attention API for LayoutLMv3 (SDPA, Flash Attn v2 support)

### LINKED ISSUE #35467: Support Flash Attention 2 for LayoutLMv3
### Feature request

Support Flash Attention 2 for LayoutLMv3.

### Motivation

Speed up document transformer training and inference.

### Your contribution

I can add it.

### PR DESCRIPTION
**Pull request by stancld**
# What does this PR do?

Closes #35467.

## Performance benchmark

Speed & memory req consumption on a token classification ntraining of LayoutLMv3-like model with multilingual support, various auxiliary tasks, masked language modelling.

GPU: 1x A100 80 GB
Batch size: 16, Accumulated gradient batches: 8

| Impl. | Speed | Peak memory |
| --- | --- | --- |
| Eager | ~2.0 it/s | 66.7 Gi |
| SDPA | ~3.0 it/s | 47.2 Gi |

Overall, ~50% speed-up and memory reqs reduction is observed.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

cc: @ArthurZucker

### CODE REVIEW

**Comment by stancld** [comment]
Just pinging @NielsRogge or @ArthurZucker .)

File: tests/models/layoutlmv3/test_modeling_layoutlmv3.py [hunk 1]
```diff
@@ -378,6 +379,28 @@ def test_model_from_pretrained(self):
...
+        tokenizer = AutoTokenizer.from_pretrained("hf-tiny-model-private/tiny-random-LayoutLMv3Model")
+        words = "I am in Paris and".split()
+        inputs = tokenizer(text=[words], boxes=[[(0, 0, 1, 1)] * len(words)], return_tensors="pt")
+
+        with torch.no_grad():
+            res_eager = model(**inputs)
+            res_sdpa = model_sdpa(**inputs)
+            self.assertTrue(
+                torch.allclose(res_eager.last_hidden_state, res_sdpa.last_hidden_state, atol=1e-5, rtol=1e-4)
+            )
```
Comment (qubvel) [inline tests/models/layoutlmv3/test_modeling_layoutlmv3.py:402]: Thanks for adding a test, but we have a common test `test_eager_matches_sdpa_inference`, so no need for another one. But it's great to see this one pass

**Review by qubvel — COMMENTED** [review]
Hi @stancld! Thanks for working on this, looks great 🤗  It would be nice to add some benchmark results to docs to ensure SDPA works after than eager one

File: docs/source/en/model_doc/layoutlmv3.md [hunk 2]
```diff
@@ -41,6 +41,26 @@ This model was contributed by [nielsr](https://huggingface.co/nielsr). The Tenso
   Due to these differences in data preprocessing, one can use [`LayoutLMv3Processor`] which internally combines a [`LayoutLMv3ImageProcessor`] (for the image modality) and a [`LayoutLMv3Tokenizer`]/[`LayoutLMv3TokenizerFast`] (for the text modality) to prepare all data for the model.
 - Regarding usage of [`LayoutLMv3Processor`], we refer to the [usage guide](layoutlmv2#usage-layoutlmv2processor) of its predecessor.
 
+### Using Scaled Dot Product Attention (SDPA)
+
+PyTorch includes a native scaled dot-product attention (SDPA) operator as part of `torch.nn.functional`. This function 
+encompasses several implementations that can be applied depending on the inputs and the hardware in use. See the 
+[official documentation](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) 
+or the [GPU Inference](https://huggingface.co/docs/transformers/main/en/perf_infer_gpu_one#pytorch-scaled-dot-product-attention)
+page for more information.
```
Comment (stevhliu) [inline docs/source/en/model_doc/layoutlmv3.md:50]:
```suggestion
PyTorch includes a native scaled dot-product attention (SDPA) operator as part of [torch.nn.functional](https://pytorch.org/docs/stable/nn.functional.html). This function 
encompasses several memory-efficient attention implementations that can be applied depending on the inputs and hardware. See the 
[official documentation](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) 
or the [GPU Inference](https://huggingface.co/docs/transformers/main/en/perf_infer_gpu_one#pytorch-scaled-dot-product-attention)
page for more information.
```

File: docs/source/en/model_doc/layoutlmv3.md [hunk 3]
```diff
@@ -41,6 +41,26 @@ This model was contributed by [nielsr](https://huggingface.co/nielsr). The Tenso
...
+`attn_implementation="sdpa"` in `from_pretrained()` to explicitly request SDPA to be used.
+
+```
+from transformers import LayoutLMv3Model
+
+model = LayoutLMv3Model.from_pretrained("microsoft/layoutlmv3-base", torch_dtype=torch.float16, attn_implementation="sdpa")
+...
+```
+
+For the best speedups, we recommend loading the model in half-precision (e.g. `torch.float16` or `torch.bfloat16`).
```
Comment (stevhliu) [inline docs/source/en/model_doc/layoutlmv3.md:62]:
```suggestion
SDPA is used by default for `torch>=2.1.1` when an implementation is available, but you may also set 
`attn_implementation="sdpa"` in `from_pretrained()` to explicitly request SDPA.

For the best speedups, we recommend loading the model in half-precision (`torch.float16` or `torch.bfloat16`).

```py
from transformers import LayoutLMv3Model

model = LayoutLMv3Model.from_pretrained("microsoft/layoutlmv3-base", torch_dtype=torch.float16, attn_implementation="sdpa")
```

**Review by stevhliu — APPROVED** [review]
Thanks!

**Review by ArthurZucker — COMMENTED** [review]
Very welcome!
Sorry for delaying the review, I saw that it was not clean (unrealted example files)
If you want to push this through let's maybe use the latest api for attention interface! (see modeling llama's attention layer!)

**Comment by stancld** [comment]
@ArthurZucker It required rebase, dunno why it looked like those unrelated files were touched.

Will check the new API.

**Comment by stancld** [comment]
@ArthurZucker Flash Attn impl is still broken here. Will have a look as time allows .]

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 4]
```diff
@@ -481,33 +449,94 @@ def forward(self, hidden_states: torch.Tensor, input_tensor: torch.Tensor) -> to
...
+        self.config = config
+        self.num_attention_heads = config.num_attention_heads
+        self.attention_head_size = config.hidden_size // config.num_attention_heads
+
+        self.self = nn.ModuleDict(
+            {
+                "query": nn.Linear(config.hidden_size, config.num_attention_heads * self.attention_head_size),
+                "key": nn.Linear(config.hidden_size, config.num_attention_heads * self.attention_head_size),
+                "value": nn.Linear(config.hidden_size, config.num_attention_heads * self.attention_head_size),
+            }
```
Comment (ArthurZucker) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py:470]: guessing we cannot remove this for BC! OK 🤗

File: src/transformers/models/layoutlmv3/modeling_layoutlmv3.py [hunk 5]
```diff
@@ -1384,6 +1430,19 @@ def forward(
         )
 
 
+def _cogview_attention(attention_scores: torch.Tensor, alpha: Union[int, float] = 32) -> torch.Tensor:
```
Comment (ArthurZucker) [inline src/transformers/models/layoutlmv3/modeling_layoutlmv3.py:1433]: we dont'really need a separate function but if we keep it place it at the top please

**Review by ArthurZucker — APPROVED** [review]
thanks
